archaius2-test/src/main/java/com/netflix/archaius/test/Archaius2TestConfig.java

```diff
@@ -35,255 +35,238 @@
  * 
  * See {@link TestPropertyOverride} for additional usage information.
  */
-public class Archaius2TestConfig implements TestRule, Config {
-    private final ThreadLocal<Config> configHolder = new ThreadLocal<>();
+public class Archaius2TestConfig implements TestRule, Config, SettableConfig {
+    private  TestCompositeConfig testCompositeConfig;
+    private final TestPropertyOverrideAnnotationReader annotationReader = new TestPropertyOverrideAnnotationReader();
 
-    public Statement apply(Statement base, Description description) {
-        return statement(base, description);
-    }
-
-    private Statement statement(final Statement base, final Description description) {
+    public Statement apply(final Statement base, final Description description) {
         return new Statement() {
             @Override
             public void evaluate() throws Throwable {
-                try {
-                    createConfig(description);
-                    base.evaluate();
-                } finally {
-                    cleanupConfig(description);
-                }
-            }
+                SettableConfig classLevelProperties = new DefaultSettableConfig();
+                SettableConfig methodLevelProperties = new DefaultSettableConfig();
+                for(Class<?> parentClass : ClassUtils.getAllSuperclasses(description.getTestClass())) {
+                    classLevelProperties.setProperties(annotationReader.getPropertiesForAnnotation(parentClass.getAnnotation(TestPropertyOverride.class)));
```

Comment:
If you have a three class hierarchy, C extends B extends A, I think that this loop will result in class level annotations being prioritized as C over A over B.  Put another way, I think that you need to process the list by getAllSuperclasses in reverse order.